Sync `RTCDataChannelInit` as per Web Specification

Ahmad-S792 opened the pull request:
#### 629c47d913460df4effc13f9c1d58333e648be90
<pre>
Sync `RTCDataChannelInit` as per Web Specification
<a href="https://bugs.webkit.org/show_bug.cgi?id=277604">https://bugs.webkit.org/show_bug.cgi?id=277604</a>
<a href="https://rdar.apple.com/problem/133630397">rdar://problem/133630397</a>

Reviewed by Youenn Fablet.

This patch aligns &apos;RTCDataChannelInit&apos; with Web Specification [1]:

[1] <a href="https://w3c.github.io/webrtc-pc/#dom-rtcdatachannelinit">https://w3c.github.io/webrtc-pc/#dom-rtcdatachannelinit</a>

It adds &apos;[EnforceRange]&apos; to &apos;maxPacketLifeTime&apos; and &apos;maxRetransmits&apos;.

Additionally, I just did fly-by fix of adding link of web
specification for easier reference.

NOTE: We match now with Gecko / Firefox as well.

* Source/WebCore/Modules/mediastream/RTCPeerConnection.idl:
* LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCDataChannelInit-maxPacketLifeTime-enforce-range-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCDataChannelInit-maxPacketLifeTime-enforce-range.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCDataChannelInit-maxRetransmits-enforce-range-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCDataChannelInit-maxRetransmits-enforce-range.html: Added.

Canonical link: <a href="https://commits.webkit.org/304499@main">https://commits.webkit.org/304499@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e1e053990102eab236d8f8b84317fcaf2e53783f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135748 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8124 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47041 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143462 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/87380 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/61a1bb5a-1fd2-4fa3-9de2-9ea32881d9a3) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/137617 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8781 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7971 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103739 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/87380 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/190063a0-1083-4fc7-bcc5-0da366a64033) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138694 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6323 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121666 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84615 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e9609d4c-27f6-4333-9462-8db5edc3e15c) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/6080 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3695 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4068 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115307 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39853 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146211 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7813 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40421 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112101 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7839 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6552 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112482 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28540 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5948 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117966 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/61746 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7859 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36073 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7594 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71408 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7820 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7691 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/mediastream/RTCPeerConnection.idl:
```diff
@@ -35,17 +35,18 @@
     boolean ordered = true;
-    unsigned short maxPacketLifeTime;
-    unsigned short maxRetransmits;
+    [EnforceRange] unsigned short maxPacketLifeTime;
```
This is web observable and requires adding a layout test.

cdumez replied:
@youennf Aligning with the specification (and Gecko) here would start throwing an exception in cases where we previously did not (which could in theory break pages/apps). Our behavior would also start diverging from Chrome. How risky is this? Should we still go ahead with this change? I am slightly worried about compatibility here so I'll defer to @youennf who is the area expert.

cdumez commented on Source/WebCore/Modules/mediastream/RTCPeerConnection.idl:
```diff
@@ -35,17 +35,18 @@
-    unsigned short maxPacketLifeTime;
-    unsigned short maxRetransmits;
+    [EnforceRange] unsigned short maxPacketLifeTime;
+    [EnforceRange] unsigned short maxRetransmits;
```
Ditto.

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCDataChannelInit-maxPacketLifeTime-enforce-range.html:
```diff
@@ -0,0 +1,117 @@
+
+test(() => {
+  const pc = createPC();
+  const channel = pc.createDataChannel('test', { maxPacketLifeTime: 1 });
```
Note that ideally the tests would be in WPT since the browsers are not aligned with the specification here and we want them to.

Ahmad-S792 replied:
Sure, if Youenn is fine with test + change, then before merging, I will move it to webrtc folder and then sync back.

youennf reviewed (approved):
Let's make sure to upload it as a WPT.
Let me know if you need a review.

youennf commented on LayoutTests/webrtc/RTCDataChannelInit-maxPacketLifeTime-enforce-range.html:
```diff
@@ -0,0 +1,117 @@
+  const pc = createPC();
+  assert_throws_js(TypeError, () => {
+    pc.createDataChannel('test', { maxPacketLifeTime: -1 });
+  });
```
I would tend to make sure that pc.close() is closed as a clean_up step.
And maybe we could have an array of bad values that we iterate and create a test.

Ahmad-S792 replied:
Good point, I will see, if I can incorporate these suggestion in test and improve it. Thanks for suggestion.

Ahmad-S792 replied:
@youennf - Updated test as WPT with some of your suggestion, does it look good, if yes, I can merge this PR and then export the test to WPT.

Commit:
Sync `RTCDataChannelInit` as per Web Specification
https://bugs.webkit.org/show_bug.cgi?id=277604
rdar://problem/133630397

Reviewed by Youenn Fablet.

This patch aligns 'RTCDataChannelInit' with Web Specification [1]:

[1] https://w3c.github.io/webrtc-pc/#dom-rtcdatachannelinit

It adds '[EnforceRange]' to 'maxPacketLifeTime' and 'maxRetransmits'.

Additionally, I just did fly-by fix of adding link of web
specification for easier reference.

NOTE: We match now with Gecko / Firefox as well.

* Source/WebCore/Modules/mediastream/RTCPeerConnection.idl:
* LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCDataChannelInit-maxPacketLifeTime-enforce-range-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCDataChannelInit-maxPacketLifeTime-enforce-range.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCDataChannelInit-maxRetransmits-enforce-range-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCDataChannelInit-maxRetransmits-enforce-range.html: Added.

Canonical link: https://commits.webkit.org/304499@main
